### PR TITLE
Add Enterprise SONIC LLDP Global module

### PR DESCRIPTION
**Pull request by divya-balasubramania**
##### SUMMARY
LLDP_GLOBAL Module implementation

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
sonic_lldp_global

##### ADDITIONAL INFORMATION

1. Regression test HTML report:  
[sonic_lldp_global_test_results.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/10848594/sonic_lldp_global_test_results.pdf)

2. Regression test detailed terminal log:
[lldp_global_terminal_test.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/10848597/lldp_global_terminal_test.txt)

3. Show running-config and respective facts gathered:
[LLDP_Global_facts_UT.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/10848601/LLDP_Global_facts_UT.txt)

### CODE REVIEW

File: plugins/modules/sonic_lldp_global.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+# -------------
+#
+# sonic# show running-configuration | grep lldp
+# sonic#
+
+  - name: Delete default LLDP configurations
+    dellemc.enterprise_sonic.sonic_lldp_global:
+      config:
+        tlv_select:
+          system_capabilities: True
```
Comment (ArunSaravananBalachandran) [inline plugins/modules/sonic_lldp_global.py:153]: Please change the value to `true`

File: plugins/modules/sonic_lldp_global.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    dellemc.enterprise_sonic.sonic_lldp_global:
+      config:
+              enable: false
+              multiplier: 9
+              system_name : CR_sonic
+              hello_time: 18
+              mode: receive
+              system_description: Sonic_System
+              tlv_select:
+                  management_address: True
```
Comment (ArunSaravananBalachandran) [inline plugins/modules/sonic_lldp_global.py:207]: Please change the value to `true`

File: tests/regression/roles/sonic_lldp_global/defaults/main.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+  - name: test_case_02
+    description: Update Global LLDP configuration
+    state: merged
+    input:
+        multiplier: 10 
+        system_name: CR1_sonic
+        hello_time: 200 
+        tlv_select: 
+           system_capabilities: true
```
Comment (ArunSaravananBalachandran) [inline tests/regression/roles/sonic_lldp_global/defaults/main.yml:29]: Please update the indentation to be uniform across blocks for all test cases.

File: tests/regression/roles/sonic_lldp_global/defaults/main.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    input:
+        enable: true
+        multiplier: 9
+        system_name: CR_sonic
+        hello_time: 18
+        mode: receive
+        system_description: Sonic_CR1_device
+        tlv_select: 
+            system_capabilities: false
+
```
Comment (ArunSaravananBalachandran) [inline tests/regression/roles/sonic_lldp_global/defaults/main.yml:19]: Please remove trailing whitespaces in this file, cleanup_tests.yaml, preparation_tests.yaml and tasks_template_del.yaml

File: plugins/module_utils/network/sonic/config/lldp_global/lldp_global.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+    def get_modify_specific_lldp_global_param_requests(self, command):
+        """Get requests to modify specific LLDP Global configurations
+        based on the command specified for the interface
+        """
+        requests = []
+
+        if not command:
+            return requests
+        if "enable" in command:
+            payload = {'openconfig-lldp:enabled': command['enable']}
```
Comment (ArunSaravananBalachandran) [inline plugins/module_utils/network/sonic/config/lldp_global/lldp_global.py:201]: Replace double quotes with single quote for keys in config and facts.

**Review by ArunSaravananBalachandran — CHANGES_REQUESTED** [review]

**Comment by divya-balasubramania** [comment]
Addressed review comments and re-executed the test cases.

Regression test HTML report:

[lldp_global_regression_results.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/10956240/regression-2023-03-13-15-19-50.pdf)

Regression test detailed terminal log:

[lldp_global_detailed_log_latest.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/10956248/lldp_global_detailed_log_latest.txt)

File: plugins/module_utils/network/sonic/facts/lldp_global/lldp_global.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                lldp_global_data['multiplier'] = raw_lldp_global_data['openconfig-lldp-ext:multiplier']
+            lldp_global_data['tlv_select'] = {}
+            lldp_global_data['tlv_select']['management_address'] = True
+            lldp_global_data['tlv_select']['system_capabilities'] = True
+            if 'suppress-tlv-advertisement' in raw_lldp_global_data:
+                for tlv_select in raw_lldp_global_data['suppress-tlv-advertisement']:
+                    tlv_select = tlv_select.replace('openconfig-lldp-types:', '').lower()
+                    if tlv_select in ('management_address', 'system_capabilities'):
+                        lldp_global_data['tlv_select'][tlv_select] = False
+        return lldp_global_data
```
Comment (ArunSaravananBalachandran) [inline plugins/module_utils/network/sonic/facts/lldp_global/lldp_global.py:113]: Default settings of 'enabled' and 'tlv_select' options needs to be handled.

File: plugins/modules/sonic_lldp_global.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+DOCUMENTATION = """
+---
+module: sonic_lldp_global
+version_added: '2.1.0'
+short_description: Manage Global LLDP configurations on SONiC
+description:
+  - This module provides configuration management of global LLDP parameters
+    for use on LLDP enabled Layer 2 interfaces of devices running SONiC.
+    It is intended for use in conjunction with LLDP Layer 2 interface
+    configuration applied on participating interfaces.
```
Comment (ArunSaravananBalachandran) [inline plugins/modules/sonic_lldp_global.py:42]: Please mention it as a list element to display it in a separate line.

File: plugins/modules/sonic_lldp_global.py
Comment (ArunSaravananBalachandran) [inline plugins/modules/sonic_lldp_global.py:1]: Please update the indentation to be uniform across blocks for all examples.

File: tests/regression/roles/sonic_lldp_global/defaults/main.yml [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+      hello_time: 200
+      tlv_select:
+        system_capabilities: true
+
+  - name: test_case_03
+    description: Delete specific global lldp configurations
+    state: deleted
+    input:
+      multiplier: 10
+      system_name: CR1_sonic
```
Comment (ArunSaravananBalachandran) [inline tests/regression/roles/sonic_lldp_global/defaults/main.yml:35]: Please include deletion of 'tlv_select' option as mentioned in the module example.

File: plugins/module_utils/network/sonic/config/lldp_global/lldp_global.py [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+            commands = have
+            requests.extend(self.get_delete_lldp_global_completely_requests(commands))
+        else:
+            if "tlv_select" in want and want['tlv_select']:
+                if "management_address" in want['tlv_select']:
+                    if want['tlv_select']['management_address'] is not None and have['tlv_select']['management_address'] is not None:
+                        want['tlv_select']['management_address'] = False
+                if "system_capabilities" in want['tlv_select']:
+                    if want['tlv_select']['system_capabilities'] is not None and have['tlv_select']['management_address'] is not None:
+                        want['tlv_select']['system_capabilities'] = False
```
[the remark this answers is not in the document]
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/lldp_global/lldp_global.py:172]: As stated in the comment (https://github.com/ansible-network/resource_module_models/pull/196#discussion_r1101081735), the deletion needs to be carried out only for matching configurations. 

Usage of diff to get the matching configuration in other options will also work for tlv_select options.
Changing the option value will result in incorrect behaviour.

File: plugins/module_utils/network/sonic/config/lldp_global/lldp_global.py [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+            payload = {'openconfig-lldp-ext:multiplier': int(command['multiplier'])}
+            url = self.lldp_global_config_path['multiplier']
+            requests.append({'path': url, 'method': PATCH, 'data': payload})
+
+        if 'system_name' in command:
+            payload = {'openconfig-lldp:system-name': command['system_name']}
+            url = self.lldp_global_config_path['system_name']
+            requests.append({'path': url, 'method': PATCH, 'data': payload})
+
+        if 'system_description' in command:
```
Comment (ArunSaravananBalachandran) [inline plugins/module_utils/network/sonic/config/lldp_global/lldp_global.py:217]: Please include `None` value check.

File: plugins/module_utils/network/sonic/config/lldp_global/lldp_global.py [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+        # The tlv_select configs are enabled by default.Hence false leads deletion of configs.
+        if 'tlv_select' in command:
+            if 'management_address' in command['tlv_select']:
+                if command['tlv_select']['management_address'] is False:
+                    url = self.lldp_suppress_tlv.format(lldp_suppress_tlv="MANAGEMENT_ADDRESS")
+                    requests.append({'path': url, 'method': DELETE})
+            if 'system_capabilities' in command['tlv_select']:
+                if command['tlv_select']['system_capabilities'] is False:
+                    url = self.lldp_suppress_tlv.format(lldp_suppress_tlv="SYSTEM_CAPABILITIES")
+                    requests.append({'path': url, 'method': DELETE})
```
Comment (ArunSaravananBalachandran) [inline plugins/module_utils/network/sonic/config/lldp_global/lldp_global.py:287]: Both `True` and `False` values need to be handled.

**Review by ArunSaravananBalachandran — CHANGES_REQUESTED** [review]

**Comment by divya-balasubramania** [comment]
Addressed review comments and re-executed the test cases.

Regression test HTML report:
[SONIC_LLDP_Global_Latest.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/11099391/SONIC_LLDP_Global_Latest.pdf)

**Review by ArunSaravananBalachandran — APPROVED** [review]

**Pull request merged**